### DISCUSSION TITLE
[운영체제] 민수 -> 진

### DISCUSSION THREAD

**Discussion by skyriv213** [tempdku/temp#25] [General]
@jjinleee 
1. Process Control Block이 문맥교환을 진행할 때 어떤 역할을 하는지 설명해주세요
2. 교착 상태 (deadlock)을 방지, 예방하기 위해 필요한 조건을 설명해주세요
3. 비선점 / 선점 스케줄링을 하나씩 예를 들고 해당 스케줄링이 각각 어떤 상황에 사용되고, 실제 해당 스케줄링을 사용해서 구현된 예를 하나씩 설명해주세요
4. 멀티 프로세스를 사용해서 구현하기에 적절한 프로그램의 예시와 멀티 스레딩을 통해 구현하기 적절한 프로그램의 예시를 설명해주세요

**Comment by jjinleee** [comment]
1. PCB는 각 프로세스를 관리하기 위한 자료구조로 문맥교환시 현재 프로세스 상태를 저장합니다. 문맥교환이 발생할시 스케줄러를 호출하여 다음 실행할 프로세스를 결정합니다. 문맥 교환 중에는 프로세스의 상태를 실행->준비, 준비-> 실행 등으로 적절히 변경합니다. 그 다음 문맥 교환 이후 실행될 프로세스의 상태를 복원합니다. 
PCB로 프로세스를 관리하고 문맥교환을 원활히 수행하도록 합니다. 이를 통해 효율적인 자원관리가 가능하고 공정한 cpu 시간 분배가 가능합니다. 때문에 다중 프로세스 환경에서 안정적으로 운영이 가능합니다.

**Reply by jjinleee** [reply]
2. 교착상태는 두개이상의 프로세스가 자원을 기다리며 무한한 대기 상태에 빠지는 것을 말합니다. 교착상태 발생 조건은 네가지가 존재합니다. 먼저 상호배제로 자원은 오직 하나의 프로세스만 사용하며 공유하지않습니다. 또한 최소 하나의 자원을 공유하고 있으며 동시에 다른 자원을 기다리는 프로세스가 있는 점유와 대기 조건이 있습니다. 세번째 조건은 이미 할당된 자원은 끝나기전까지 강제적으로 뺏을수없는 비선점 조건입니다. 마지막 조건은 순환대기입니다. 
교착상태를 방지하기 위해서 이 네가지 조건을 만족하지 않도록 하면 됩니다. 

- 상호배제조건 제거 
- 점유와 대기 조건제거 (프로세스가 자원을 요청할시 현재 보유자원이 없다면 요청 거절)
- 비선점 조건 제거 (자원점유중인 프로세스가 다른 자원요청시 보유한 자원을 반납하도록)
- 순환대기조건제거( 자원에 고유순서를 정하고 프로세스가 요청시 정한 순서대로만 요청하도록)

이 외에도 교착상태를  방지하기 위한 은행가 알고리즘이 존재합니다. 이는 프로세스가 자원을 요청할때 시스템이 이요청을 승인하더라도 안전한 상태인지 확인하고 안전상태일 경우에만 자원요청을 승인합니다.

**Comment by jjinleee** [comment]
3. 비선점 스케줄링의 예시로 SJF(shortest job first)가 있습니다.  이는 대기열에 있는 프로세스 중에서 실행시간이 가장 짧은 프로세스를 먼저 실행하는 스케줄링 알고리즘입니다. 이는 주로 배치 처리 시스템에서 사용됩니다. 이것이 구현된 예로는 프린트 서버가 대표적입니다. 프린트 서버는 가장 짧은 인쇄작업을 우선처리하여 대기시간을 줄입니다. 
선점 스케줄링의 예시로 우선순위 스케줄링이 있습니다. 각 프로세스에 우선순위를 부여하고 높은 우선순위의 프로세스 먼저 실행하는 형식입니다. 실행중인 프로세스보다 높은 우선순위를 가진 프로세스가 도착하면 실행중인 프로세스는 중단되고 새로운 프로세스가 실행됩니다.  이것이 구현된 예로는 RTOS(real-time operating system) 가 있습니다. 이는 실시간 응답이 필요한 시스템에서 중요한 작업을 처리합니다.  사용예시로는 항공기 제어 시스템과 산업용 로봇 시스템 센서입력에 사용됩니다.

**Comment by jjinleee** [comment]
4. 멀티 프로세싱은 여러 개의 프로세스를 사용하여 병렬 처리를 수행하는 방식으로 데이터 과학 및 분석 작업에 용이합니다. 왜냐하면 대량의 데이터를 처리하는 과정이 필요하기 때문에 성능을 향상시킬 수 있습니다. 
멀티 스레딩은 하나의 프로세스내에 여러개의 스레드를 사용하여 병렬처리를 수행하는 방식으로  다수의 클라이언트 요청을 처리하는 웹 서버나 채팅 프로그램에 용이합니다.
